internal/bminventory/inventory_v2_handlers.go

```diff
@@ -605,13 +606,63 @@ func (b *bareMetalInventory) signURL(ctx context.Context, infraEnvID, urlString,
 	return urlString, &expiresAt, nil
 }
 
-const ipxeScriptFormat = `#!ipxe
+const ipxeFirstScriptFormat = `#!ipxe
+chain %s&mac=${net0/mac}
+`
+
+const ipxeSecondScriptFormat = `#!ipxe
 initrd --name initrd %s
 kernel %s initrd=initrd coreos.live.rootfs_url=%s random.trust_cpu=on rd.luks.options=discard ignition.firstboot ignition.platform.id=metal console=tty1 console=ttyS1,115200n8 coreos.inst.persistent-kargs="console=tty1 console=ttyS1,115200n8"
 boot
 `
 
-func (b *bareMetalInventory) infraEnvIPXEScript(ctx context.Context, infraEnv *common.InfraEnv) (string, error) {
+func (b *bareMetalInventory) firstIPXEScript(infraEnv *common.InfraEnv) (string, error) {
+	serviceBaseURL := b.ServiceBaseURL
+	if b.InsecureServiceBaseURL != "" {
+		serviceBaseURL = b.InsecureServiceBaseURL
+	}
+	parsedURL, err := url.Parse(serviceBaseURL)
+	if err != nil {
+		return "", err
+	}
+	builder := installer.V2DownloadInfraEnvFilesURL{
+		InfraEnvID: *infraEnv.ID,
+		FileName:   "ipxe-script",
+	}
+	return fmt.Sprintf(ipxeFirstScriptFormat, builder.StringFull(parsedURL.Scheme, parsedURL.Host)), nil
+}
+
+func (b *bareMetalInventory) shouldServeScript(infraEnv *common.InfraEnv, mac *strfmt.MAC) (bool, error) {
+	var hosts []*models.Host
+	macStr := mac.String()
+	if err := b.db.Where("infra_env_id = ? and (inventory like ? or inventory like ?)", infraEnv.ID.String(), fmt.Sprintf("%%%s%%", strings.ToUpper(macStr)),
+		fmt.Sprintf("%%%s%%", strings.ToLower(macStr))).Find(&hosts).Error; err != nil {
+		return false, err
+	}
+	switch len(hosts) {
+	case 0:
+		return true, nil
+	case 1:
+		break
+	default:
+		return false, common.NewApiError(http.DefaultMaxHeaderBytes, errors.Errorf("Unexpected number of hosts %d with mac %s", len(hosts), macStr))
+	}
+	h := hosts[0]
+	switch swag.StringValue(h.Status) {
+	case models.HostStatusInstalled:
+		return false, nil
+	case models.HostStatusInstallingInProgress:
+		if h.Progress != nil {
+			switch h.Progress.CurrentStage {
+			case models.HostStageDone, models.HostStageConfiguring, models.HostStageJoined, models.HostStageRebooting, models.HostStageWaitingForIgnition:
+				return false, nil
```

Comment:
Since a `false` here will cause an API error I think it would be great to get some information about why we're erroring in the response and in the log.

Do you think we could refactor this function so that it only returns an error type and we serve the script only when it succeeds?

Reply:
If we do this refactoring the function might need another rename too.

Reply:
I will do the refactor. Look at line 701.  An error is generated with specific text.  Anything else is needed ?

Reply:
Yeah, I think it would be good to create error text that includes _exactly_ why we're skipping the script.
Like the current status/stage if that's the reason or the mac we're searching for if we couldn't find the host.

I'm imagining a scenario where a user isn't getting the script served to them and doesn't know why. The easiest way to debug this would be to have this information both in the log and the response text.